internal/app/synchronizer/synchronizer_service.go

```diff
@@ -16,145 +16,541 @@ package synchronizer
+		// Wait for ignore list, but not all matches returned, the next cycle
+		// can start now.
+		close(closedOnCycleEnd)
+	}()
 
-		proposals = append(proposals, req.GetMatch())
+	closeRegistration := time.After(s.registrationInterval())
+Registration:
+	for {
+		select {
+		case req := <-s.synchronizeRegistration:
+			allM1cSent.Add(1)
+			callingCtx = append(callingCtx, req.ctx)
+			r := &registration{
+				m1c:        m1c,
+				m6c:        make(chan *pb.Match),
+				cancelMmfs: make(chan struct{}, 1),
+				cycleCtx:   ctx,
+				allM1cSent: &allM1cSent,
+			}
+			registrations = append(registrations, r)
+			req.resp <- r
+		case <-closeRegistration:
+			break Registration
+		}
 	}
 
-	results, err := s.doEvaluateProposals(stream.Context(), proposals, id)
-	if err != nil {
-		return err
+	go func() {
+		for _, ctx := range callingCtx {
+			<-ctx.Done()
+		}
+		cancel(fmt.Errorf("Canceled because all callers were done."))
+	}()
+
+	go func() {
+		allM1cSent.Wait()
+		m1c.cutoff()
+	}()
+
+	cancelProposalCollection := time.AfterFunc(s.proposalCollectionInterval(), func() {
+		m1c.cutoff()
+		for _, r := range registrations {
+			r.cancelMmfs <- struct{}{}
+		}
+	})
+	<-closedOnCycleEnd
+
+	// Clean up in case it was never needed.
+	cancelProposalCollection.Stop()
+}
+
+type registrationRequest struct {
+	resp chan *registration
+	ctx  context.Context
+}
+
+type registration struct {
+	m1c        *cutoffSender
+	allM1cSent *sync.WaitGroup
+	m6c        chan *pb.Match
+	cancelMmfs chan struct{}
+	cycleCtx   context.Context
+}
+
+///////////////////////////////////////
+///////////////////////////////////////
+
+type mAndM6c struct {
+	m   *pb.Match
+	m6c chan *pb.Match
+}
+
+// For incoming mathces, remembers the channel the match needs to be returned to
+// if it passes the synchronizer.
+func fanInFanOut(m2c <-chan mAndM6c, m3c chan<- *pb.Match, m5c <-chan *pb.Match) {
+	m6cMap := make(map[string]chan<- *pb.Match)
+
+	defer func() {
+		for range m2c {
+		}
+	}()
+
+loop:
+	for {
+		select {
+		case m2, ok := <-m2c:
+			if !ok {
+				close(m3c)
+				break loop
+			}
+			m6cMap[m2.m.GetMatchId()] = m2.m6c
+			m3c <- m2.m
+
+		case m5, ok := <-m5c:
+			if !ok {
+				return
+			}
+
+			m6cMap[m5.GetMatchId()] <- m5
+		}
 	}
 
-	for _, result := range results {
-		if err = stream.Send(&ipb.EvaluateProposalsResponse{Match: result}); err != nil {
-			return err
+	for m5 := range m5c {
+		m6cMap[m5.GetMatchId()] <- m5
+	}
+}
+
+///////////////////////////////////////
+///////////////////////////////////////
+
+type cutoffSender struct {
+	m1c       chan<- mAndM6c
+	m2c       chan<- mAndM6c
+	closed    chan struct{}
+	closeOnce sync.Once
+}
+
+// cutoffSender allows values to be passed on the provided channel until cutoff
+// has been called.  This closed the provided channel.  Calls to send after
+// cutoff work, but values are ignored.
+func newCutoffSender(m2c chan<- mAndM6c) *cutoffSender {
```

Comment:
Ah I see this is an effort to avoid panics from closing a closed channel since we may have multiple go routines writing to the same one.

Does this trick fit our use case at here? It might be able to make the code more straightforward if so.
https://stackoverflow.com/a/39237154

Reply:
No, because changing the channel's value requires editing the variable on multiple go routines.